wandb/sdk/internal/context.py

```diff
@@ -0,0 +1,79 @@
+"""Context Keeper."""
+
+import threading
+from typing import Dict, Optional
+
+from wandb.proto.wandb_internal_pb2 import Record, Result
+
+
+class Context:
+    _cancel_event: threading.Event
+    _debug_record: Optional[Record]
+
+    def __init__(self) -> None:
+        self._cancel_event = threading.Event()
+        self._debug_record = None
+
+    def cancel(self) -> None:
+        self._cancel_event.set()
+
+    @property
+    def cancel_event(self) -> threading.Event:
+        return self._cancel_event
+
+
+def context_id_from_record(record: Record) -> str:
+    context_id = record.control.mailbox_slot
+    return context_id
+
+
+def context_id_from_result(result: Result) -> str:
+    context_id = result.control.mailbox_slot
+    return context_id
+
+
+class ContextKeeper:
```

Comment:
totally nit: should we rename it?
- i would replace the keeper with tracker 
- i think the context noun is confusing (association of context manager, but we are actually just keep track of mailbox slots), but i don't have something better for this